### PR TITLE
allow changing song

### PR DESCRIPTION
**Pull request by willmtemple**
Signed-off-by: William Temple <willmtemple@gmail.com>

### CODE REVIEW

File: bot.py [hunk 1]
```diff
@@ -2,6 +2,14 @@
 import discord
 import asyncio
```
Comment (smmalis37) [inline bot.py:4]: Remove the blank line.

File: bot.py [hunk 2]
```diff
@@ -13,6 +21,10 @@
 colors = list(filter(lambda x: isinstance(discord_colors[x],classmethod), discord_colors))
 colors.sort()
 
+async def get_video_title(url):
+    yt = etree.HTML(urllib.urlopen(url).read())
+    return ' '.join(yt.xpath("//span[@id='eow-title']/@title"))
```
Comment (smmalis37) [inline bot.py]: Is there no way to do this without adding a dependency?

File: bot.py [hunk 3]
```diff
@@ -32,6 +44,20 @@
             await client.replace_roles(author, *new_roles)
             await client.send_message(message.channel, "Changed your color to " + words[1])
 
+async def change_song(message):
+    global current_track
+    global elevator
+    try:
+        song = message.content.split(' ')[1]
+    except IndexError:
```
Comment (smmalis37) [inline bot.py]: Do an if len instead of catching an error.

File: bot.py [hunk 4]
```diff
@@ -97,18 +126,20 @@ def is_friend(member):
...
     global in_elevator
+    global current_track
     voice_channel = discord.utils.get(after.server.channels, name=elevatorName)
     if in_elevator and after.id == client.user.id and after.voice.voice_channel != voice_channel:
         await client.move_member(after, voice_channel)
     elif in_elevator and len(voice_channel.voice_members) == 1:
         elevator.stop()
         elevator = None
         await client.voice_client_in(after.server).disconnect()
+        current_track = ELEVATOR_MUSIC
```
Comment (smmalis37) [inline bot.py:129]: This seems like an odd decision. If we've changed the song wouldn't we want it to stay changed?

File: bot.py [hunk 5]
```diff
@@ -82,6 +108,9 @@ def is_friend(member):
                 
         if command == '!color':
             await handle_color(message)
+
+        if command == '!song':
```
Comment (smmalis37) [inline bot.py:104]: Should this be an admin only command? It seems open to abuse.

File: bot.py [hunk 6]
```diff
@@ -32,6 +44,20 @@
             await client.replace_roles(author, *new_roles)
             await client.send_message(message.channel, "Changed your color to " + words[1])
 
+async def change_song(message):
+    global current_track
+    global elevator
+    try:
+        song = message.content.split(' ')[1]
+    except IndexError:
+        print('The current song is : {}'.format(get_video_title(current_track)))
```
Comment (smmalis37) [inline bot.py]: This won't send to the discord channel, it'll print to the console. Is that intentional?

**Review by smmalis37 — CHANGES_REQUESTED** [review]

File: bot.py (hunk 4 above)
Comment (a-scales) [reply bot.py:129]: I agree on this. I would rather have the new song than the old one.

Comment (willmtemple) [reply bot.py:129]: no

File: bot.py (hunk 5 above)
Comment (willmtemple) [reply bot.py:104]: no

File: bot.py (hunk 6 above)
Comment (willmtemple) [reply bot.py]: no

File: bot.py (hunk 2 above)
Comment (willmtemple) [reply bot.py]: no

File: bot.py (hunk 1 above)
Comment (willmtemple) [reply bot.py:4]: ok

File: bot.py (hunk 3 above)
Comment (willmtemple) [reply bot.py]: ok

File: bot.py (hunk 2 above)
Comment (smmalis37) [reply bot.py]: Can't ytdl_player do this for us? Or would that cause us to download the video itself twice?

File: bot.py [hunk 7]
```diff
@@ -32,6 +36,21 @@
...
             await client.send_message(message.channel, "Changed your color to " + words[1])
 
+async def change_song(message):
+    global current_track
+    global elevator
+    contents = message.content.split(' ')
+    if len(contents) == 2
+        song = message.content.split(' ')[1]
+    else:
+        client.send_message(message.channel, 'The current song is : "{}"'.format(elevator.title))
```
Comment (smmalis37) [inline bot.py]: This needs to await.

Comment (smmalis37) [reply bot.py]: And probably should return, since song won't be set otherwise.

**Review by smmalis37 — CHANGES_REQUESTED** [review]

File: bot.py (hunk 4 above)
Comment (willmtemple) [reply bot.py:129]: @a-scales this is not a discussion

File: bot.py [hunk 8]
```diff
@@ -32,6 +36,21 @@
...
             await client.send_message(message.channel, "Changed your color to " + words[1])
 
+async def change_song(message):
+    global current_track
+    global elevator
+    contents = message.content.split(' ')
+    if len(contents) == 2
+        song = message.content.split(' ')[1]
+    else:
+        await client.send_message(message.channel, 'The current song is : "{}"'.format(elevator.title))
```
Comment (smmalis37) [inline bot.py]: This else should return, since song won't be set.

**Review by smmalis37 — CHANGES_REQUESTED** [review]

File: bot.py [hunk 9]
```diff
@@ -82,6 +102,9 @@ def is_friend(member):
                 
         if command == '!color':
             await handle_color(message)
+
+        if command == '!song':
```
Comment (smmalis37) [inline bot.py:104]: Add this command to !help

File: bot.py [hunk 10]
```diff
@@ -32,6 +36,22 @@
...
+async def change_song(message):
+    global current_track
+    global elevator
+    contents = message.content.split(' ')
+    if len(contents) == 2
+        song = message.content.split(' ')[1]
+    else:
+        await client.send_message(message.channel, 'The current song is : "{}"'.format(elevator.title))
+        return
+    current_track = song
```
Comment (smmalis37) [inline bot.py]: Move all this code into the if

**Review by smmalis37 — CHANGES_REQUESTED** [review]

**Review by smmalis37 — APPROVED** [review]

**Pull request merged**